### ISSUE TITLE
_has_bits_ was not declared in this scope

### ISSUE THREAD

**Issue by jwatte** [facebookarchive/caffe2#1195]
Building on ubuntu 17.04 with GCC 5.4.1 and protoc version 3.0.0.

I have previously successfully checked out and built caffe2 on this system.

Today I did a fresh `git clone --recursive github.com:caffe2/caffe2` followed by a `sudo ./scripts/build_local.sh` and after building protoc successfully, the caffe2 build fails with almost every file saying:

    /usr/local/src/caffe2/build/caffe/proto/caffe.pb.h: In member function ‘bool caffe::V0LayerParameter::has_det_fg_fraction() const’:
    /usr/local/src/caffe2/build/caffe/proto/caffe.pb.h:25250:11: error: ‘_has_bits_’ was not declared in this scope
       return (_has_bits_[0] & 0x10000000u) != 0;

I think the problem is that "build_host_protoc" despite calling `make install` (and being run with sudo) doesn't actually install the binaries from `build_host_protoc/bin/`

**Comment by jwatte** [comment]
I've been trying to build the latest trunk on ubuntu 17.04 / x86_64 for a while, and found out a bunch of things.

  * The std::string ABI changed between gcc4 and gcc5.
  * protobuf uses std::string in its ABI.
  * The scripts/build_local.sh script defines a custom protobuf executable
  * However, the cmake files look for the protobuf that's installed on the machine for linux builds
  * The default protobuf on ubuntu, which is required for the ubuntu-desktop to run, is 3.0.0
  * The custom protobuf executable defined in scripts/build_local.sh is the locally built version, which is 3.1.0
  * When building local protobuf without shared libraries, it builds without -fPIC, which generates symbols that are not relocatable in shared libraries
  * The python plugins are of course shared libraries and require relocatable symbols
  * The python plugins will attempt to link against the system protobuf libraries because that's the default ldconfig search path, unless I link them statically

I think the best option is to make build_local.sh specify static protobuf libraries, but edit the cmake files (all of them) to make sure -fPIC is included in the compile options.
Also, update ProtoBuf.cmake to always call the custom protobuf find function, so that we don't use a protoc that mis-matches the system libraries. Always use the built-in version, and always statically link.

Also:

  * The protobuf test files use the deprecated "tmpnam" library function (that's been deprecated for years.)
  * The caffe2 headers fail compile in client applications with -Wall -Werror because they contain inline functions that compare signed values to unsigned values.

**Comment by jwatte** [comment]
Or perhaps simpler: Remove the custom PROTOC command definition from the build_local.sh script.

**Comment by jwatte** [comment]
Attaching a patch.

**Comment by jwatte** [comment]
```
diff --git a/scripts/build_local.sh b/scripts/build_local.sh
index 4e0d62b7..84ccd953 100755
--- a/scripts/build_local.sh
+++ b/scripts/build_local.sh
@@ -16,15 +16,19 @@ mkdir -p $BUILD_ROOT
 echo "Build Caffe2 into: $BUILD_ROOT"
 
 # Build protobuf from third_party so we have a host protoc binary.
-echo "Building protoc"
-$CAFFE2_ROOT/scripts/build_host_protoc.sh || exit 1
+PROTOC_DEFS=""
+if [ ! -x /usr/bin/protoc ]; then
+    echo "Building protoc"
+    $CAFFE2_ROOT/scripts/build_host_protoc.sh || exit 1
+    PROTOC_DEFS="-DPROTOBUF_PROTOC_EXECUTABLE=$CAFFE2_ROOT/build_host_protoc/bin/protoc"
+fi
 
 # Now, actually build the target.
 echo "Building caffe2"
 cd $BUILD_ROOT
 
 cmake .. \
-    -DPROTOBUF_PROTOC_EXECUTABLE=$CAFFE2_ROOT/build_host_protoc/bin/protoc \
+    $PROTOC_DEFS \
     -DBUILD_SHARED_LIBS=OFF \
     || exit 1
-make
+make -j 16

```

**Comment by pietern** [comment]
Thanks for your investigation here @jwatte. I made #1328 to address the mismatch problem.

**Comment by pietern** [comment]
It also addresses the PIC issue when using the static libprotobuf bits built as part of the C2 bits by setting the CMake PIC property on the libprotobuf target. I saw those relocation errors as well.

**Comment by pietern** [comment]
Closing as this is addressed in #1328.
